경북대 BE 김진현: Step 1 - 카카오 로그인

kimjhyun0627 opened the pull request:
## 구현한 기능

- 카카오 애플리케이션 등록: 이전에 다른 백엔드 프레임워크에서 사용한 경험이 있어 오래 걸리지는 않았습니다.
- 환경변수 등록: `application.properties`에 토큰을 등록하고 `.gitignore`에 등록했습니다.
- 외부 Client와 DTO 구현: Kakao Auth Server로의 요청을 위해 Client를 구현하고, 이 때 사용할 DTO를 구현했습니다.
- 서비스 레이어 구현: 기존 `MemberService`에 메소드를 추가했습니다.
- 컨트롤러 구현: 기존 `MemberController`에 엔드포인트를 추가했습니다.
- Client와 Service, Controller에 대한 테스트 코드를 추가했습니다.

### 질문
- 현재 Member DB가 `email`, 해시처리된 `passwordhash`를 통해 새 유저를 받고 있기에, 카카오 로그인한 계정을 등록하는 과정이 쉽지 않았던 것 같습니다(카카오 로그인 시에 email과 password 항목을 함께 받아올 수 없었습니다 ㅜㅜ)
- 차선책으로, 이메일 필드를 `kakao_{사용자ID}@noemail.kakao`로 두고, 해시처리된 비밀번호 필드 역시 {사용자ID}를 해시처리하여 DB에 저장하고 있습니다.(도메인 제약조건으로 `nullable=false`를 해제하는 것이 더 위험하다고 판단했습니다!)
- 단기적으로는 괜찮지만, 장기적으로는 문제가 발생할 여지가 있을것 같은데, 이런 경우 따로 서비스 DB에 등록하지 않고 처리하는 것이 맞는지, 혹은 DB 필드를 추가하는 방법을 고민해봐야하는지 여쭤보고 싶습니다!

kimjhyun0627 committed:
feat: 카카오 애플리케이션 등록 & 환경변수 등록

kimjhyun0627 committed:
feat: DTO & 외부 Client 구현

kimjhyun0627 committed:
feat: 서비스 레이어 구현

kimjhyun0627 committed:
feat: 컨트롤러 구현

woodying commented on .gitignore:
```diff
@@ -5,6 +5,9 @@ build/
 !**/src/test/**/build/
 
+# 민감정보 분리
+**/application.properties
```
민감정보 분리를 위해 `application.properties` 설정을 전부 노출하지 않도록 하는 것 보다는, 민감 정보만 분리하는게 좋아보입니다.
`application-secrets.properties` 에 민감정보를 분리해두고, `spring.profiles.include=secrets` 설정을 통해 import 하는 방법이 있습니다~ 🙂

woodying commented on src/main/java/gift/external/kakao/dto/KakaoTokenResponse.java:
```diff
@@ -0,0 +1,16 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record KakaoTokenResponse(
+        @JsonProperty("access_token")
```
snake_case 로 jackson 매핑을 하고싶으신거라면, class 레벨에 아래와 같은 설정을 통해 가능합니다~

`@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)`

woodying commented on src/main/java/gift/repository/member/MemberRepository.java:
```diff
@@ -1,10 +1,12 @@
 
 public interface MemberRepository extends JpaRepository<Member, Long> {
 
     Optional<Member> findByEmail_Email(String email);
```
함수명이 이상합니다~
`findByEmail` 로 수정해주세요 🙂

woodying commented on src/test/java/gift/service/member/MemberServiceTest.java:
```diff
@@ -213,4 +216,58 @@ void deleteMemberAsUserThrows() {
+    @DisplayName("Kakao 로그인")
+    class KakaoLoginTests {
+
+        @Mock
+        KakaoTokenClient kakaoTokenClient;
```
mock 잘 활용해주셨네요 👍👍

woodying commented on src/main/java/gift/config/RestTemplateConfig.java:
```diff
@@ -0,0 +1,14 @@
+@Configuration
+public class RestTemplateConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
+        return new RestTemplate();
+    }
```
`RestTemplate` 을 통해 외부 클라이언트와 통신하는것으로 보이는데요~
설정에 connect/read timeout 에 대한 설정이 보이지 않습니다. 

외부 클라이언트와 통신에 문제가 있는 경우, 무한대로 상대를 기다리게 되면 서비스 전체의 문제로 확산될 수 있습니다 🥲
이러한 설정은 명시적으로 추가해주셨으면 좋겠습니다. 🙂

kimjhyun0627 replied:
`SimpleClientHttpRequestFactory` 사용해 connect/read 타임아웃 추가했습니다!
놓치고 있었네요;) 감사합니다 !

woodying commented on src/main/java/gift/service/member/MemberServiceImpl.java:
```diff
@@ -100,6 +108,28 @@ public void deleteMember(Long id, Role role) {
+        accessToken = kakaoTokenClient.getAccessToken(authorizationCode);
+        userInfo = kakaoTokenClient.getUserInfo(accessToken);
+
+        String kakaoId = String.valueOf(userInfo.id());
+        String email = "kakao_" + kakaoId + "@noemail.kakao";
+
+        Member member = memberRepository.findByEmail_Email(email)
+                .orElseGet(() -> {
+                    Member newMember = Member.register(email, sha256(email));
+                    return memberRepository.save(newMember);
+                });
```
여기서 email 과 register() 에 사용하는 임의변조 값들은 다소 어색하네요 😅
보통 이런식으로 사용하지 않습니다.

kimjhyun0627 replied:
`ProviderType`과 `OauthId` 필드를 도메인에 추가해서 조금 더 자연스럽게 소셜 회원가입을 진행할 수 있도록 수정했습니다 !

woodying reviewed (changes requested):
진현님 안녕하세요~ 🙂

이전에 다른 프로젝트에서 소셜로그인 기능을 연동해보신 적이 있으시군요. 
코드를 살펴봤는데, 소셜로그인 후 사용자를 등록하는 부분을 제외하면 대부분 잘 구현해주신 것 같아요. 👍👍
다만, connect/read timeout 에 대한 설정은 추가해주셨으면 좋겠어요~ 
외부 클라이언트가 응답하지 않을때 무한정 기다릴 순 없으니까요. 🙃

고민하셨던 부분에 대해서 말씀드리겠습니다. 
먼저, 소셜로그인으로 로그인한 유저도 물론 `member` 테이블에 등록하고 관리해주는 것이 좋습니다. 🙂 이 때 방법은 정책에 따라 조금 달라질 수도 있을 것 같습니다.
`이메일이 같으면 같은유저로 볼것이냐` or `이메일이 같아도 로그인 방법이 다르면 다른 유저로 볼것이냐` 에 따라서 다르게 처리할 수 있어 보이네요.
또, member의 password 값은 nullable 이 맞다고 생각합니다. 😊
소셜로그인으로 로그인했다면 password 값은 없을테니까요~ 
다만, 카카오에서 발급받은 유저에 대한 유니크 키값(app_user_id/이메일 등)을 저장해두는 방법이 있겠네요. 
그리고 회원테이블에 providerType 같은 필드도 추가하고 활용하면 좋을 것 같습니다. (id/pw로그인 or 소셜로그인)

한 번 고민해보시고 수정 부탁드립니다~ 
진행하시다가 어려운점 있으면 말씀부탁드려요! 🙂

kimjhyun0627 committed:
refactor: step 1 코드리뷰 반영

- application.properties 분리
- KakaoTokenResponse: Jackson 매핑
- MemberRepository.findByEmail: 메소드명 변경
- RestTemplate: connect/read 타임아웃 설정
- schema.sql: member 스키마 변경
- Member: 도메인 변경 & password nullable
- MemberService: oauth 로그인에 대해 password=null, oauthId 속성 추가

kimjhyun0627 commented:
리뷰 감사합니다! 말씀해주신 부분 수정해 보았습니다.

주요 변경점은
- 사용자 등록 로직 변경: 도메인에 `ProviderType`, `OauthId` 필드를 추가하고, `PasswordHash`를 `nullable`로 변경했습니다.
   - 기존 회원가입 시에는 이메일과 비밀번호로 회원가입을 할 수 있습니다(`ProviderType=NONE`)
   - 카카오 로그인 시에는 `kakaoId`로 목업 이메일을 만들어서 이메일 필드를 채우고, password는 `null`, `ProviderType=KAKAO`, `OauthId={카카오ID}`로 회원 정보를 DB에 추가하였습니다.
- `RestTemplate`의 connect/read 시 타임아웃을 설정했습니다.

woodying commented on .gitignore:
```diff
@@ -5,6 +5,9 @@ build/
 !**/src/test/**/build/
 
+# 민감정보 분리
+**/application-secrets.properties
```
👍

woodying commented on src/main/java/gift/config/RestTemplateConfig.java:
```diff
@@ -0,0 +1,19 @@
+    @Bean
+    public RestTemplate restTemplate() {
+        SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+        requestFactory.setConnectTimeout(5000);
+        requestFactory.setReadTimeout(3000);
```
타임아웃 설정 추가해주셨네요 👍👍

woodying commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -37,6 +38,14 @@ public class Member {
     @Column(name = "created_at", nullable = false, updatable = false)
     private LocalDateTime createdAt;
 
+    @Enumerated(EnumType.STRING)
+    @Column(name = "provider_type", nullable = false)
+    private ProviderType providerType;
```
👍

woodying commented on src/main/java/gift/entity/member/Member.java:
```diff
@@ -46,24 +55,48 @@ private Member(
         );
     }
 
+    public static Member registerOauth(
```
👍

woodying commented on src/main/java/gift/service/member/MemberServiceImpl.java:
```diff
@@ -100,6 +110,27 @@ public void deleteMember(Long id, Role role) {
+        userInfo = kakaoTokenClient.getUserInfo(accessToken);
+
+        String kakaoId = String.valueOf(userInfo.id());
+        String email = "kakao_user_" + kakaoId + "@oauth.local"; // or @fake.email
+
+        Member member = memberRepository.findByOauthIdAndProviderType(kakaoId, ProviderType.KAKAO)
+                .orElseGet(() -> {
+                    Member newMember = Member.registerOauth(kakaoId, ProviderType.KAKAO, email);
+                    return memberRepository.save(newMember);
+                });
```
email 도 일반적인 소셜로그인 후 받는 이메일 그대로 저장할 수 있을 것 같습니다.
만약, 소셜로그인 후 이메일을 얻을 수 없거나하면 email 도 역시 nullable 이 맞는 것 같습니다~ 🙂
나중에 설계해야한다면 이 부분도 고려해주세요~!
큰 문제는 없어서 당장 수정할 필요는 없어보입니다 👍

woodying commented on src/main/java/gift/external/kakao/dto/KakaoTokenResponse.java:
```diff
@@ -0,0 +1,19 @@
+        @JsonProperty("access_token")
+        String accessToken,
+
+        @JsonProperty("token_type")
```
`@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)` 을 설정해주셔서, 필드에서는 이 설정을 제거해도됩니다! 🙂

woodying commented on src/main/java/gift/controller/user/AuthController.java:
```diff
@@ -39,4 +40,12 @@ public ResponseEntity<AuthResponse> login(
         return ResponseEntity.ok(authResponse);
     }
+
+    @PostMapping("login/kakao")
```
이 method 는 `GET` 이 되어야 하지 않나요?
front 에서 카카오 로그인 후 인가코드를 들고 redirect url 로 `GET` 요청을 할 것 같습니다.
한 번 확인부탁드려요 🙂

kimjhyun0627 replied:
`@GetMapping`으로 수정했습니다 !

woodying reviewed (approved):
진현님 안녕하세요~ 

1단계 진행해주시느라 고생많으셨습니다. 
피드백 드렸던 부분들도 빠르게 반영해주셨네요 👍👍

크게 더 고쳐야할 부분은 없는 것 같아서 1단계는 마무리하겠습니다. 🙂
코멘트 남긴 부분들은 한 번 읽어보시고, 가능한 부분들은 2단계에 함께 반영부탁드려요~

다음 단계도 파이팅입니다~! 👏
